ray-operator/controllers/ray/rayservice_controller.go

```diff
@@ -72,3 +158,203 @@ func (r *RayServiceReconciler) SetupWithManager(mgr ctrl.Manager) error {
+			return nil, err
+		}
+		if err := r.Create(context.TODO(), rayClusterInstance); err != nil {
+			r.Log.Error(err, "unable to create rayCluster for rayService", "rayCluster", rayClusterInstance)
+			// Error creating the RayCluster object - requeue the request.
+			return nil, err
+		}
+		r.Log.V(1).Info("created rayCluster for rayService run", "rayCluster", rayClusterInstance)
+	} else {
+		r.Log.Error(err, "Get request rayCluster instance error!")
+		// Error reading the RayCluster object - requeue the request.
+		return nil, client.IgnoreNotFound(err)
+	}
+
+	return rayClusterInstance, nil
+}
+
+func (r *RayServiceReconciler) constructRayClusterForRayService(rayService *rayv1alpha1.RayService) (*rayv1alpha1.RayCluster, error) {
+	name := fmt.Sprintf("%s%s", rayService.Name, rayClusterSuffix)
+
+	rayCluster := &rayv1alpha1.RayCluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Labels:      rayService.Labels,
+			Annotations: rayService.Annotations,
+			Name:        name,
+			Namespace:   rayService.Namespace,
+		},
+		Spec: *rayService.Spec.RayClusterSpec.DeepCopy(),
+	}
+
+	// Set the ownership in order to do the garbage collection by k8s.
+	if err := ctrl.SetControllerReference(rayService, rayCluster, r.Scheme); err != nil {
+		return nil, err
+	}
+
+	return rayCluster, nil
+}
+
+func (r *RayServiceReconciler) fetchDashboardURL(rayCluster *rayv1alpha1.RayCluster) (string, error) {
+	headServices := corev1.ServiceList{}
+	filterLabels := client.MatchingLabels{common.RayClusterLabelKey: rayCluster.Name}
+	if err := r.List(context.TODO(), &headServices, client.InNamespace(rayCluster.Namespace), filterLabels); err != nil {
+		return "", err
+	}
+
+	dashboardURL := ""
+
+	if headServices.Items != nil {
+		if len(headServices.Items) == 1 {
+			r.Log.Info("reconcileServices ", "head service found", headServices.Items[0].Name)
+			// TODO: compare diff and reconcile the object. For example. ServiceType might be changed or port might be modified
+			servicePorts := headServices.Items[0].Spec.Ports
+
+			dashboardPort := int32(-1)
+
+			for _, servicePort := range servicePorts {
+				if servicePort.Name == "dashboard" {
+					dashboardPort = servicePort.Port
+					break
+				}
+			}
+
+			if dashboardPort == int32(-1) {
+				return "", fmtErrors.Errorf("dashboard port not found")
+			}
+
+			dashboardURL = fmt.Sprintf("%s.%s.svc.cluster.local:%v",
+				headServices.Items[0].Name,
+				headServices.Items[0].Namespace,
+				dashboardPort)
+			return dashboardURL, nil
+		}
+
+		// This should never happen.
+		// We add the protection here just in case controller has race issue or user manually create service with same label.
+		if len(headServices.Items) > 1 {
+			r.Log.Info("reconcileServices ", "Duplicates head service found", len(headServices.Items))
+			return "", fmtErrors.Errorf("Duplicates head service found %v", len(headServices.Items))
+		}
+	}
+
+	// Create head service if there's no existing one in the cluster.
```

Comment:
The code doesn't match the comment at the moment.
We're relying on the RayCluster reconciler reconciler to create the head service, right?